src/boot/efi/boot.c

```diff
@@ -1871,6 +1872,115 @@ static bool is_sd_boot(EFI_FILE *root_dir, const char16_t *loader_path) {
         return memcmp(content, magic, sizeof(magic)) == 0;
 }
 
+static EFI_STATUS launch_internal_uefi_shell(const ConfigEntry *entry) {
+        EFI_STATUS err;
+
+        VENDOR_DEVICE_PATH *dp;
+        err = BS->HandleProtocol(entry->device, MAKE_GUID_PTR(EFI_DEVICE_PATH_PROTOCOL), (void **) &dp);
+        if (err != EFI_SUCCESS)
+                return err;
+
+        if (dp->Header.Type != MEDIA_DEVICE_PATH || dp->Header.SubType != MEDIA_PIWG_FW_VOL_DP)
+                return EFI_INVALID_PARAMETER;
+
+        static struct {
+                VENDOR_DEVICE_PATH fv;
+                VENDOR_DEVICE_PATH ff;
+                EFI_DEVICE_PATH end;
+        } _packed_ internal_uefi_shell_dp = {
+                .fv = {
+                        .Header = {
+                                .Type = MEDIA_FILEPATH_DP,
+                                .SubType = MEDIA_PIWG_FW_VOL_DP,
+                                .Length = sizeof(internal_uefi_shell_dp.fv),
+                        },
+                },
+                .ff = {
+                        .Header = {
+                                .Type = MEDIA_FILEPATH_DP,
+                                .SubType = MEDIA_PIWG_FW_FILE_DP,
+                                .Length = sizeof(internal_uefi_shell_dp.ff),
+                        },
+                        .Guid = UEFI_SHELL_GUID,
+                },
+                .end = DEVICE_PATH_END_NODE,
+        };
+
+        internal_uefi_shell_dp.fv.Guid = dp->Guid;
+
+        _cleanup_(unload_imagep) EFI_HANDLE shell;
+        err = BS->LoadImage(false, IMG, (EFI_DEVICE_PATH *) &internal_uefi_shell_dp, NULL, 0, &shell);
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Error loading internal UEFI shell: %m");
+
+        err = BS->StartImage(shell, NULL, NULL);
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Error starting internal UEFI shell: %m");
+
+        return EFI_SUCCESS;
+}
+
+static void config_entry_add_loader_uefi_internal(Config *config) {
```

Comment:
i don't like the name "internal" for this. internal suggests it was internal to *us*. but it's internal to the ed2k firmware, not us, hence the name is a bit misleading.

hence, maybe name it "config_entry_add_firmware_shell()"? then it's pretty clear, no?